src/server/cmd/pachctl/shell/completions.go

```diff
@@ -215,6 +217,50 @@ func PipelineCompletion(_, _ string, maxCompletions int64) ([]prompt.Suggest, Ca
 	return result, CacheAll
 }
 
+func jobSetDesc(jsi *pps.JobSetInfo) string {
+	failure := 0
+	var created *types.Timestamp
+	var modified *types.Timestamp
+	for _, job := range jsi.Jobs {
+		if job.State != pps.JobState_JOB_SUCCESS && ppsutil.IsTerminal(job.State) {
+			failure++
+		}
+
+		if created == nil {
+			created = job.Created
+			modified = job.Created
+		} else {
+			if job.Created.Compare(created) < 0 {
+				created = job.Created
+			}
+			if job.Created.Compare(modified) > 0 {
+				modified = job.Created
+			}
+		}
+	}
+	return fmt.Sprintf("%s, %d subjobs %d failure(s)", pretty.Ago(created), len(jsi.Jobs), failure)
+}
+
+// JobCompletion completes job parameters of the form <job-set>
+func JobSetCompletion(_, text string, maxCompletions int64) ([]prompt.Suggest, CacheFunc) {
+	c := getPachClient()
+	var result []prompt.Suggest
+	listJobSetClient, err := c.PpsAPIClient.ListJobSet(c.Ctx(), &pps.ListJobSetRequest{})
```

Comment:
If this is for autocomplete, should this context have a timeout? 3-5 seconds maybe?

Reply:
I believe that's typically set by the caller. I could maybe see the argument that the completion layer should set its own lower timeout so as not to block things for too long. Because if the completion function blocks then the shell just locks up. But it's a little hard to know what the right timeout would be and I think what we have now has been working reasonably well so I'm inclined to just leave it.